contracts/src/BoundlessMarketCallback.sol

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) 2025 RISC Zero, Inc.
+//
+// All rights reserved.
+pragma solidity ^0.8.24;
+
+import {IRiscZeroVerifier, Receipt, ReceiptClaim, ReceiptClaimLib} from "risc0/IRiscZeroVerifier.sol";
+import {IBoundlessMarketCallback} from "./IBoundlessMarketCallback.sol";
+
+/// @notice Contract for handling proofs delivered by the Boundless Market's callback mechanism.
+/// @dev This contract provides a framework for applications to safely handle proofs delivered by
+/// the Boundless Market for a specific image ID. The intention is for developers to inherit the contract
+/// and implement the internal `_handleProof` function.
+/// @dev We recommend a best practice of "trust but verify" whenever receiving proofs, so we verify the proof
+/// here even though the Boundless Market already verifies the proof as part of its fulfillment process.
+/// Proof verification in Boundless is cheap as it is just a merkle proof, so this adds minimal overhead.
+abstract contract BoundlessMarketCallback is IBoundlessMarketCallback {
+    using ReceiptClaimLib for ReceiptClaim;
+
+    IRiscZeroVerifier public immutable VERIFIER;
+    address public immutable BOUNDLESS_MARKET;
+    bytes32 public immutable IMAGE_ID;
+
+    /// @notice Initializes the callback contract with verifier and market addresses
+    /// @param verifier The RISC Zero verifier contract address
+    /// @param boundlessMarket The BoundlessMarket contract address
+    /// @param imageId The image ID to accept proofs of.
+    constructor(IRiscZeroVerifier verifier, address boundlessMarket, bytes32 imageId) {
+        VERIFIER = verifier;
+        BOUNDLESS_MARKET = boundlessMarket;
+        IMAGE_ID = imageId;
+    }
+
+    /// @inheritdoc IBoundlessMarketCallback
+    function handleProof(bytes32 imageId, bytes calldata journal, bytes calldata seal) public {
+        require(msg.sender == BOUNDLESS_MARKET, "Invalid sender");
```

Comment:
I think this is fine to include, and possibly beneficial, but its not clear to me that this is case. Let's keep this check for now, and clearly folks can always override this function to remove it if they decide it's unnecessary.

Keep in mind that anyone can specify a callback to any contract, and anyone can create a request on the market. So the amount of constraint this applies to the call is limited, beyond the fact that the receipt is already checked, which is rechecked based on our discussion earlier.